### PR TITLE
Cleaned up recursive script loading and fixed script loading on mac

### LINKED ISSUE #968: Fix Script Loading on Mac
Script loading should have more states than passed or not. Also, there are problems with loading scripts on mac. (claimed)

### PR DESCRIPTION
**Pull request by MatthewFrench**
Resolve #968

### CODE REVIEW

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs [hunk 1]
```diff
@@ -32,13 +40,24 @@ public bool LoadSubdirectoryScripts(string folder)
...
+                return CompilationStatus.NoScripts;
+            }
+            return Load(new List<string>(allfiles));
         }
 
-        //Takes about 300 milliseconds for a single script
-        public bool Load(List<string> scriptLocations)
+        /// <summary>
+        /// Loads scripts from a list of files and compiles them.
+        /// Takes about 300 milliseconds for a single script. Faster for a banch of scripts.
```
Comment (lzardy) [inline GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs]: typo

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs [hunk 2]
```diff
@@ -32,13 +40,24 @@ public bool LoadSubdirectoryScripts(string folder)
...
+        /// Returns true for success.
+        /// </summary>
+        /// <param name="scriptLocations"></param>
+        /// <returns>True means success</returns>
```
Comment (lzardy) [inline GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs]: What about false?

Comment (MatthewFrench) [reply GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs]: I changed it to reflect the enum that gets returned. Good catch on making sure my documentation was accurate.

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs [hunk 3]
```diff
@@ -53,50 +72,58 @@ public bool Load(List<string> scriptLocations)
             var assemblyName = Path.GetRandomFileName();
 
             var references = new List<MetadataReference>();
-            foreach (var a in AppDomain.CurrentDomain.GetAssemblies())
-                if (!a.IsDynamic && !a.Location.Equals(""))
-                    references.Add(MetadataReference.CreateFromFile(a.Location));
+            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
```
Comment (lzardy) [inline GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:77]: Have you tested performance on both for and foreach?

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs [hunk 4]
```diff
@@ -53,50 +72,58 @@ public bool Load(List<string> scriptLocations)
...
+            {
+                if (!assembly.IsDynamic && !assembly.Location.Equals(""))
+                {
+                    references.Add(MetadataReference.CreateFromFile(assembly.Location));
+                }
+            }
+            //Add vector2 reference because vector2 isn't in System.Numerics in dot net core
```
Comment (lzardy) [inline GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:84]: This isn't true for either [.NET Core](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.vector2?view=netcore-3.1), nor [.NET Framework ](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.vector2?view=netframework-4.8).

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs (hunk 3 above)
Comment (MatthewFrench) [reply GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:77]: I don't think for loop performance is necessary here. This should only be called once on game start or when manually reloading all scripts.
For is said to be 2-5x faster but sacrifices in readability. There won't be a perceivable difference here in performance.

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs (hunk 4 above)
Comment (MatthewFrench) [reply GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:84]: It unfortunately is true for .net core on mac. At least on our current version of .net core which is 3.0.
https://github.com/dotnet/standard/issues/783
Vector2 is in its own nuget package: https://www.nuget.org/packages/System.Numerics.Vectors/

Without this line, scripts do not compile with the error of: 

> The type name 'Vector2' could not be found in the namespace 'System.Numerics'. This type has been forwarded to assembly 'System.Numerics.Vectors, Version=4.1.5.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' Consider adding a reference to that assembly.

So the line I added, adds a reference to that assembly.

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs [hunk 5]
```diff
@@ -53,50 +72,58 @@ public bool Load(List<string> scriptLocations)
...
+            references.Add(MetadataReference.CreateFromFile(typeof(Vector2).Assembly.Location));
             //Now add game reference
             references.Add(MetadataReference.CreateFromFile(typeof(Game).Assembly.Location));
-            var op = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary)
+            var compilationOptions = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary)
                 .WithOptimizationLevel(OptimizationLevel.Release).WithConcurrentBuild(true);
```
Comment (lzardy) [inline GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:89]: Unsure if this has been a hindrance, but it can be addressed in this PR so may as well.

Comment (MatthewFrench) [reply GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:89]: What are you referring to as a hindrance?

Comment (MatthewFrench) [reply GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:89]: If you mean the name change, acronyms and abbreviations should be avoided wherever possible. I like to clean up files I work in.

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs [hunk 6]
```diff
@@ -53,50 +72,58 @@ public bool Load(List<string> scriptLocations)
...
-                    errored |= true;
-                    var invalidSourceTrees = GetInvalidSourceTrees(result);
-
-                    if (invalidSourceTrees.Count == 0)
+                    else
                     {
-                        // Shouldnt happen
-                        _logger.Error("Script compilation failed");
-                        return true;
+                        state = CompilationStatus.SomeCompiled;
```
Comment (lzardy) [inline GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:118]: Does nothing?

Comment (MatthewFrench) [reply GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:118]: That changes the state from everything compiled to only some things compiled. This enum is checked outside this class to print out a message stating only some of the scripts compiled.
You can test that this works by introducing an error in one script and seeing the console output that only some of the scripts compiled.

Comment (MatthewFrench) [reply GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs:118]: It gets returned in the next while loop run if the compilation is successful.

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs [hunk 7]
```diff
@@ -23,22 +31,38 @@ public CSharpScriptEngine()
...
             {
                 var fileBasePath = Path.GetFullPath(pathString);
                 var trimmedPath = fileBasePath.Remove(0, basePath.Length);
                 var directories = trimmedPath.ToLower().Split(Path.DirectorySeparatorChar);
-                if (directories.Contains("bin") || directories.Contains("obj")) return false;
+                if (directories.Contains("bin") || directories.Contains("obj"))
+                {
+                    return false;
+                }
+                if (pathString.Contains("AssemblyInfo.cs")) {
```
Comment (danil179) [inline GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs]: Move the bracket to a separate line.

**Review by danil179 — CHANGES_REQUESTED** [review]
Code style problem.

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs [hunk 8]
```diff
@@ -23,22 +31,39 @@ public CSharpScriptEngine()
...
             {
                 var fileBasePath = Path.GetFullPath(pathString);
                 var trimmedPath = fileBasePath.Remove(0, basePath.Length);
                 var directories = trimmedPath.ToLower().Split(Path.DirectorySeparatorChar);
-                if (directories.Contains("bin") || directories.Contains("obj")) return false;
+                if (directories.Contains("bin") || directories.Contains("obj"))
+                {
+                    return false;
+                }
+                if (pathString.Contains("AssemblyInfo.cs"))
```
Comment (danil179) [inline GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs]: Why not adding it to the previous `if`?

**Review by danil179 — APPROVED** [review]
Seems OK. Though I have a comment about merging one of the ifs. Please rebuild.

File: GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs (hunk 8 above)
Comment (MatthewFrench) [reply GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs]: I wasn't happy with the length of the line, I'll combine them.

**Pull request merged**